File: src/core/validators/AbstractEntity/AbstractEntityMustContainAnIdentity.js
// @flow
import { getProperty } from '../ValidationHelper';
import type SymbolTable from '../SymbolTable';
import { errorRuleBase } from '../ValidationRuleBase';
import { includeRuleBase } from '../ValidationRuleRepository';
import { MetaEdGrammar } from '../../../../src/grammar/gen/MetaEdGrammar';

// eslint-disable-next-line no-unused-vars
export function valid(ruleContext: any, symbolTable: SymbolTable): boolean {
  return ruleContext.property().some(x => getProperty(x).propertyComponents().propertyAnnotation().identity() != null);
}

// eslint-disable-next-line no-unused-vars
function failureMessage(ruleContext: any, symbolTable: SymbolTable): string {
  return `Abstract Entity ${ruleContext.abstractEntityName().ID().getText()} does not have an identity specified.`;
}

const validationRule = errorRuleBase(valid, failureMessage);
// eslint-disable-next-line import/prefer-default-export
export const includeRule = includeRuleBase(MetaEdGrammar.RULE_abstractEntity, validationRule);

File: src/core/validators/AbstractEntity/AbstractEntityMustContainAnIdentity.cs
using System.Linq;
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator.AbstractEntity
{
    public class AbstractEntityMustContainAnIdentity : ValidationRuleBase<MetaEdGrammar.AbstractEntityContext>
    {
        public override bool IsValid(MetaEdGrammar.AbstractEntityContext context)
        {
            return context.property().Any(x => x.GetProperty().propertyComponents().propertyAnnotation().identity() != null);
        }

        public override string GetFailureMessage(MetaEdGrammar.AbstractEntityContext context)
        {
            return string.Format("Abstract Entity {0} does not have an identity specified.", context.abstractEntityName().ID().GetText());
        }
    }
}
